crates/task-impls/src/rewind.rs

```diff
@@ -0,0 +1,81 @@
+use std::collections::VecDeque;
+use std::sync::Arc;
+
+use anyhow::Result;
+use async_broadcast::{Receiver, Sender};
+use async_trait::async_trait;
+use hotshot_task::task::TaskState;
+use hotshot_types::traits::node_implementation::NodeType;
+use std::fs::OpenOptions;
+use std::io::Write;
+
+use crate::events::HotShotEvent;
+
+#[cfg(not(any(debug_assertions, test)))]
+const REWIND_MAX_DEPTH: usize = 1_000;
+
+/// We want the depth to be huge for tests because we don't care that much about a long-running
+/// memory leak
+#[cfg(any(debug_assertions, test))]
+const REWIND_MAX_DEPTH: usize = 100_000;
+
+/// The task state for the `Rewind` task is used to capture all events received
+/// by a particular node, in the order they've been received.
+pub struct RewindTaskState<TYPES: NodeType> {
+    /// All events received by this node since the beginning of time.
+    pub events: VecDeque<Arc<HotShotEvent<TYPES>>>,
+
+    /// The id of this node
+    pub id: u64,
+}
+
+impl<TYPES: NodeType> RewindTaskState<TYPES> {
+    /// Handles all events, storing them to the private state
+    pub fn handle(&mut self, event: Arc<HotShotEvent<TYPES>>) {
+        if self.events.len() == REWIND_MAX_DEPTH {
+            self.events.pop_front();
+        }
+        self.events.push_back(Arc::clone(&event));
```

Comment:
I fear this will kind of defeat the point in production,  Lets say an issue happens (e.g. a leader doesn't propose) and the critical part was in view 1000.  We probably won't shut down the node for 1000 more views (and a view has many events) so the interesting sequence will just be lost.

I think this comes down to not waiting until shutdown until we write the events to file.

Reply:
Done